### PR TITLE
Add some automatic device protocol detection

### PR DESCRIPTION
**Pull request by mjg59**
Some devices require an 8-byte protocol (currently referred to as Ledenet),
while some require a 7-byte protocol but don't set the rgb/white flag. A
subset of the latter type support true rgbw operation. Merge the command
sending into a single function that handles this, and set appropriate flags
when determining the initial device state.

### CODE REVIEW

**Comment by Danielhiversen** [comment]
This is great.
I will not be able to test it before begging of next week

**Comment by Danielhiversen** [comment]
@bah2830. Could you test if this code works for you? Also try it without setting the protocol?

**Comment by Danielhiversen** [comment]
@tchellomello Have you tested this?

**Comment by tchellomello** [comment]
@Danielhiversen  I'll test it tonight

**Comment by mjg59** [comment]
@tchellomello Did you get a chance to test?

**Comment by Danielhiversen** [comment]
Sorry for late response :(
I have now added it to my calendar for testing on Saturday.

**Comment by tchellomello** [comment]
@mjg59  Sorry I couldn't test it yet.  I'll try to do it by the end of this week.

File: flux_led/__main__.py [hunk 1]
```diff
@@ -661,21 +660,33 @@ def getRgbw(self):
...
         msg.append(int(b))
         msg.append(int(w))
-
         if self.protocol == "LEDENET":
             msg.append(0x00)
 
-        msg.append(0x0f)
+        if not self.rgbwprotocol:
+            if w > 0:
+                bytes.append(0xf0)
```
Comment (Danielhiversen) [inline flux_led/__main__.py]: msg.append

File: flux_led/__main__.py [hunk 2]
```diff
@@ -661,21 +660,33 @@ def getRgbw(self):
...
+            else:
+                bytes.append(0x0f)
```
Comment (Danielhiversen) [inline flux_led/__main__.py]: msg.append

**Comment by Danielhiversen** [comment]
It is not working here.  I can turn the lights on/off but changing the brightness is not working. Change color gives only rely dark color,

My lights are detected as
```
        self.protocol = None
        self.rgbwcapable = False
        self.rgbwprotocol = False
```
For that case the same format as before should be used.


I might have some more time for debugging later to day.

**Comment by outrun0506** [comment]
He Guys,

just give it a try. Unfortunately my v3 Controller from LEDENET ist not working with this changes.

**Comment by mjg59** [comment]
@Danielhiversen Gah, sorry - does it work with those changes?

**Comment by Danielhiversen** [comment]
No, the behavior I describe is with these changes

**Comment by mjg59** [comment]
@Danielhiversen Hm, ok. That is odd. What calls are you making?

**Comment by Danielhiversen** [comment]
I used it with Home-assistant: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/light/flux_led.py

So
self._bulb.setWarmWhite255(brightness)
self._bulb.setRgb(*tuple(rgb))


I will try to debug some more tomorrow.

**Comment by mjg59** [comment]
@Danielhiversen Oh bother - try self.setRgbw(0, 0, 0, level, brightness=None, persist=persist, retry=retry) at the bottom of setWarmWhite255, and try naming the parameters in the call at the end of setRgb() as well?

**Comment by Danielhiversen** [comment]
It works for me with the change I add.

File: flux_led/__main__.py [hunk 3]
```diff
@@ -661,28 +661,41 @@ def getRgbw(self):
...
         else:
             msg = bytearray([0x41])
+
         msg.append(int(r))
         msg.append(int(g))
         msg.append(int(b))
         msg.append(int(w))
-
         if self.protocol == "LEDENET":
             msg.append(0x00)
```
Comment (Danielhiversen) [inline flux_led/__main__.py]: Should probably add 
```
       if level > 0
           msg.append(int(level))
      else:
           msg.append(0x00)
```
https://github.com/Danielhiversen/flux_led/pull/23/files#diff-432c4d3e79417f8676dc71ae5e419702L643

**Comment by Danielhiversen** [comment]
@bah2830  Could you also check this code?

File: flux_led/__main__.py (hunk 3 above)
Comment (mjg59) [inline flux_led/__main__.py]: Unclear on this - when could level be less than 0?

**Comment by mjg59** [comment]
Ok, looks like I may have had the white/rgb values inverted. That should be fixed now.

File: flux_led/__main__.py (hunk 3 above)
Comment (Danielhiversen) [inline flux_led/__main__.py]: Yeah, you right.
So just change msg.append(0x00)  to  msg.append(int(level)) ?

Comment (mjg59) [inline flux_led/__main__.py]: Done

**Comment by Danielhiversen** [comment]
Looks good to me. Great work.

I will wait a few more day before merging, so @tchellomello and @bah2830  get a bit more time to test.

@outrun0506 could you test the latest code? It should now work with LEDENET

**Comment by tchellomello** [comment]
Hello guys, 
   
   It worked for me here with the latest changes too. 
  Good work guys!

**Comment by outrun0506** [comment]
@Danielhiversen I'd give it try unfortunately only turning on and off is working for my v3 Controller. I should get some more time on weekend to test. I'll try to check code, if i can find it once more why my controller is not working (probably should change my controller :-) )

**Comment by Yungling** [comment]
@Danielhiversen I would like to help testing this too. Just i really have no clue  how i can do it. Can you let me know how so i can contribute a bit? Thx!

**Pull request merged**